vrpn_Freespace.C

```diff
@@ -1,55 +1,72 @@
+#if FREESPACE_VERSION_GREATER_THAN(0,7,0)
+#include <freespace/freespace_util.h>
+#endif
+
 #include <cstring>
-static const q_type gs_qIdent = Q_ID_QUAT;
 
-// We've not yet initialized the FreeSpace library.
-bool vrpn_Freespace::_freespace_initialized = false;
+static q_type gs_qIdent = Q_ID_QUAT;
 
-// libfreespace provides linear acceleration values in (mm/(s^2))^-3 
+// some libfreespace devices provide linear acceleration values in (mm/(s^2))^-3 
 // VRPN wants these in meters/(s^2)
 static vrpn_float64 convertFreespaceLinearAccelation(int16_t mmpss) {
-	return mmpss / ((vrpn_float64) 1000);
+    return mmpss / ((vrpn_float64) 1000);
 }
-// freespace reports in milliradians per second, so need to convert
+
+// some freespace devices report in milliradians per second, so need to convert
 // to meters per second...
 // The radius to use is a bit arbitrary, and would depend on 
 // what exactly this device is 'attached' to.  Additionally, this may 
 // be different for each axis.  In general, the radius shouldn't really
 // matter since the device may be representing a larger/smaller object in
 // the virtual world.
 static vrpn_float64 convertFreespaceAngularVelocity(int16_t mrps, double radius) {
-	return mrps * radius / ((vrpn_float64) 1000);
+    return mrps * radius / ((vrpn_float64) 1000);
 }
+
+// some freespace devices report in radians per second 
+static vrpn_float64 convertFreespaceAngularVelocityRadPerSecond(int16_t mrps, double radius) {
+    return mrps * radius;
+}
+
+
 #define MILLIS_TO_SECONDS(x) (x / ((vrpn_float64) 1000000))
 
+#define SCOOP_PID 0xC0B3
+#define FSM6_PID 0xC080
+#define FSM9_PID 0xC0E0
 
 vrpn_Freespace::vrpn_Freespace(FreespaceDeviceId freespaceId,
         struct FreespaceDeviceInfo* deviceInfo,
-        const char *name, vrpn_Connection *conn):
-        vrpn_Tracker_Server(name, conn),
-        vrpn_Button_Filter(name, conn),
-        vrpn_Dial(name, conn)
+        const char *name, vrpn_Connection *conn) :
+    vrpn_Tracker_Server(name, conn),
+    vrpn_Button_Filter(name, conn),
+    vrpn_Dial(name, conn),
+    _freespaceDevice(freespaceId)
 {
-	memcpy(&_deviceInfo, deviceInfo, sizeof(_deviceInfo));
-	// 5 buttons + a scroll wheel
-	vrpn_Button::num_buttons = 5;
-	vrpn_Dial::num_dials = 1;
-	memset(&_lastBodyFrameTime, 0, sizeof(_lastBodyFrameTime));
-	_timestamp.tv_sec = 0;
+    memcpy(&_deviceInfo, deviceInfo, sizeof(_deviceInfo));
+    // 5 buttons + a scroll wheel
+    vrpn_Button::num_buttons = 5;
+    vrpn_Dial::num_dials = 1;
+    memset(&_lastBodyFrameTime, 0, sizeof(_lastBodyFrameTime));
+    _timestamp.tv_sec = 0;
 }
 
 void vrpn_Freespace::freespaceInit() {
-	if (!_freespace_initialized) {
-		_freespace_initialized = true;
-		int rc = freespace_init();
-		if (rc != FREESPACE_SUCCESS) {
-			fprintf(stderr, "vrpn_Freespace::freespaceInit: failed to init freespace lib. rc=%d\n", rc);
-			_freespace_initialized = false;
-		}
-	}
+    static bool freespaceInit = false;
```

Comment:
Hiding this static variable down inside a function is a bit more subtle than having it in the main class declaration.  I think having it in the header file is better, even though it is perhaps a larger scope than needed, because a scan of the header files for "static" will make it show up.  We've done this in the past when looking for how to make things more thread safe.